### ISSUE TITLE
does django-redis support retry on TimeOut Error while reading data?

### ISSUE THREAD

**Issue by prasad5141** [jazzband/django-redis#667]
I am using django-redis package version (4.12.1),

I am getting a TimeOut Exception(Intermittent issue) while reading data from the redis cache, I would like to Retry for some N number of time to fetch the data

I tried using these settings but retry is not happening      "REDIS_CLIENT_KWARGS": {"retry_on_timeout": True}

Please let me know if any other settings are available to retry while reading data from redis using django-redis package.

**Comment by barkhabol** [comment]
@prasad5141 Were you able to execure the retry logic for n number of times by any chance?  Thanks in advance.

**Comment by esseti** [comment]
Is there any news on how to use the retry?

**Comment by WisdomPill** [comment]
sorry for the late reply, retry is not supported

**Comment by prasad5141** [comment]
Hi @WisdomPill ,

Thanks for your reply, Shall I implement Retry?

Plan of Implementation:
- User will specify the number of times to retry while reading data from Redis. (it will be in settings.py)
- Code should retry a user-specified number of times, if still not able to get the data after the specified count will throw an error `max retry exceeded`

Please suggest to me if you have any other plan for retry, and I will try to implement it.

**Comment by WisdomPill** [comment]
okay, but by default the number of retries should be zero to not change the default behaviour

**Comment by esseti** [comment]
any update? @prasad5141

**Comment by amingilani** [comment]
@esseti Mind if I take this over? I have some void time coming up and would like to contribute to this package :)

**Comment by amingilani** [comment]
I'd love some clarifications:

> Redis client supports 3 parameters to configure the retry behaviour:
>
>    retry: [Retry](https://redis-py.readthedocs.io/en/stable/retry.html#redis.retry.Retry) instance with a [Backoff](https://redis-py.readthedocs.io/en/stable/backoff.html#backoff-label) strategy and the max number of retries
>
>    retry_on_error: list of [Exceptions](https://redis-py.readthedocs.io/en/stable/exceptions.html#exceptions-label) to retry on
>
>    retry_on_timeout: if True, retry on [TimeoutError](https://redis-py.readthedocs.io/en/stable/exceptions.html#redis.exceptions.TimeoutError) only

Does this signature sound fine? I can create a `Retry` class along with a few `Backoff` classes?

**Comment by esseti** [comment]
I just saw this: where you able to implement it ?

**Comment by amingilani** [comment]
@esseti I haven't implemented it, I'm still waiting on clarification for how the interface should look like. Let me know when you can please.

**Comment by esseti** [comment]
Not sure what i can/cannot confirm, but just mimic the parameters of the client should work. that is:

- Retry: int: default 1
- RetryBackoff: one of the https://redis-py.readthedocs.io/en/stable/backoff.html#backoff-label or the default: NoBackoff()
- retry_on_error: bool
- retry_on_timeout: bool

**Comment by WisdomPill** [comment]
it is something that has to be setup at the connection factory level, following the [documentation of redis-py](https://redis-py.readthedocs.io/en/stable/retry.html#retry-in-redis-standalone) it seems to me that it can be done by adding the right parameters [here](https://github.com/jazzband/django-redis/blob/8ec90cc1f24c815e0c1069cbcd577a1d98cb6b9f/django_redis/pool.py#L30).

So I changed my mind, but I didn't try it yet, I think it can be already done using some configuration like the following

```
retry = Retry(ExponentialBackoff(), 3)
retry_on_error=[BusyLoadingError, ConnectionError, TimeoutError]
CACHES = {
    "default": {
        "BACKEND": "django_redis.cache.RedisCache",
        "LOCATION": "redis://127.0.0.1:6379/1",
        "OPTIONS": {
            "REDIS_CLIENT_KWARGS": {
               "retry": retry
               "retry_on_error": retry_on_error
            }
        }
    }
}
```

or for timeouts only

```
retry = Retry(ExponentialBackoff(), 3)
CACHES = {
    "default": {
        "BACKEND": "django_redis.cache.RedisCache",
        "LOCATION": "redis://127.0.0.1:6379/1",
        "OPTIONS": {
            "REDIS_CLIENT_KWARGS": {
               "retry": retry
               "retry_on_timeout": True
            }
        }
    }
}
```

to make it easier and cleaner we could override `REDIS_CLIENT_KWARGS ` based on `OPTIONS` like @esseti suggested.

to start @amingilani I would try to write some tests mimicking the [redis-py ones](https://github.com/redis/redis-py/blob/master/tests/test_retry.py)

**Comment by akshetpandey** [comment]
PS: This already works, but the documentation is wrong. You have to use CONNECTION_POOL_KWARGS instead of REDIS_CLIENT_KWARGS.
